lib/AST/PluginLoader.cpp

```diff
@@ -0,0 +1,150 @@
+  for (auto &arg : Ctx.SearchPathOpts.getCompilerPluginExecutablePaths()) {
+    // Create a moduleName -> pluginPath mapping.
+    assert(!arg.ExecutablePath.empty() && "empty plugin path");
+    StringRef pathStr = Ctx.AllocateCopy(arg.ExecutablePath);
+    for (auto moduleName : arg.ModuleNames) {
+      ExecutablePluginPaths[Ctx.getIdentifier(moduleName)] = pathStr;
+    }
+  }
+}
+
+void PluginLoader::setRegistry(PluginRegistry *newValue) {
+  assert(Registry == nullptr && "Too late to set a new plugin registry");
+  Registry = newValue;
+}
+
+PluginRegistry *PluginLoader::getRegistry() {
+  // Create a new one if it hasn't been set.
+  if (!Registry) {
+    Registry = new PluginRegistry();
+    OwnedRegistry.reset(Registry);
+  }
+
+  assert(Registry != nullptr);
+  return Registry;
+}
+
+llvm::Optional<std::string>
+PluginLoader::lookupLibraryPluginByModuleName(Identifier moduleName) {
+  auto fs = Ctx.SourceMgr.getFileSystem();
+
+  // Look for 'lib${module name}(.dylib|.so)'.
+  SmallString<64> expectedBasename;
+  expectedBasename.append("lib");
+  expectedBasename.append(moduleName.str());
+  expectedBasename.append(LTDL_SHLIB_EXT);
+
+  // Try '-load-plugin-library'.
+  for (const auto &libPath :
+       Ctx.SearchPathOpts.getCompilerPluginLibraryPaths()) {
+    if (llvm::sys::path::filename(libPath) == expectedBasename) {
+      return libPath;
+    }
+  }
+
+  // Try '-plugin-path'.
+  for (const auto &searchPath : Ctx.SearchPathOpts.PluginSearchPaths) {
+    SmallString<128> fullPath(searchPath);
+    llvm::sys::path::append(fullPath, expectedBasename);
+    if (fs->exists(fullPath)) {
+      return std::string(fullPath);
+    }
+  }
+
+  return None;
+}
+
+Optional<std::pair<std::string, std::string>>
+PluginLoader::lookupExternalLibraryPluginByModuleName(Identifier moduleName) {
+  auto fs = Ctx.SourceMgr.getFileSystem();
+
+  for (auto &pair : Ctx.SearchPathOpts.ExternalPluginSearchPaths) {
+    SmallString<128> fullPath(pair.SearchPath);
+    llvm::sys::path::append(fullPath,
+                            "lib" + moduleName.str() + LTDL_SHLIB_EXT);
+
+    if (fs->exists(fullPath)) {
+      return {{std::string(fullPath), pair.ServerPath}};
+    }
+  }
+  return None;
+}
+
+Optional<StringRef>
+PluginLoader::lookupExecutablePluginByModuleName(Identifier moduleName) {
+  auto &execPluginPaths = ExecutablePluginPaths;
+  auto found = execPluginPaths.find(moduleName);
+  if (found == execPluginPaths.end())
+    return None;
+  return found->second;
+}
+
+LoadedLibraryPlugin *PluginLoader::loadLibraryPlugin(StringRef path) {
+  auto fs = Ctx.SourceMgr.getFileSystem();
+  SmallString<128> resolvedPath;
+  if (auto err = fs->getRealPath(path, resolvedPath)) {
+    Ctx.Diags.diagnose(SourceLoc(), diag::compiler_plugin_not_loaded, path,
+                       err.message());
+    return nullptr;
+  }
+
+  // Track the dependency.
+  if (DepTracker)
+    DepTracker->addDependency(resolvedPath, /*IsSystem=*/false);
```

Comment:
Should we ever care to pass `IsSystem=true` here? I could imagine it being true for plugins loaded from `lib/swift/host/plugins`, but (1) I don't know if that will have any practical effect, and (2) that likely means we'd need to have both `-system-plugin-path` and `-plugin-path`, and that just feels annoying.

Reply:
> This looks great! Am I understanding correctly that this means we will rebuild every source file in the module when one of the macro implementations that the module uses will change, vs. only rebuilding those source files that use one of the macros?
> 
> Does there need to be a corresponding driver change to deal with these new dependencies, or is this all we need?

@rintaro could we please add a test to examine the `.swiftdeps ` produced when we have a non-primary source input use a plugin? Once this lands in a snapshot toolchain, would be great to also add a test in the driver incremental build test suite to validate the answer to Doug's question.  

On the surface, it looks like `.swiftdeps` may contain dependencies on plugins from non-primary sources so this may cause coarse-grained incremental re-compilation, but I think that's an okay starting point. And I am pretty sure the driver should be able to handle this without any changes for now.